src/sparse_blas/backends/cusparse/cusparse_task.hpp

```diff
@@ -0,0 +1,187 @@
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef _ONEMKL_SPARSE_BLAS_BACKENDS_CUSPARSE_TASKS_HPP_
+#define _ONEMKL_SPARSE_BLAS_BACKENDS_CUSPARSE_TASKS_HPP_
+
+#include "cusparse_handles.hpp"
+#include "cusparse_scope_handle.hpp"
+
+/// This file provide a helper function to submit host_task using buffers or USM seamlessly
+
+namespace oneapi::mkl::sparse::cusparse {
+
+template <typename T, typename Container>
+auto get_value_accessor(sycl::handler &cgh, Container container) {
+    auto buffer_ptr =
+        reinterpret_cast<sycl::buffer<T, 1> *>(container->value_container.buffer_ptr.get());
+    return buffer_ptr->template get_access<sycl::access::mode::read_write>(cgh);
+}
+
+template <typename T, typename... Ts>
+auto get_fp_accessors(sycl::handler &cgh, Ts... containers) {
+    return std::array<sycl::accessor<T, 1>, sizeof...(containers)>{ get_value_accessor<T>(
+        cgh, containers)... };
+}
+
+template <typename T>
+auto get_row_accessor(sycl::handler &cgh, matrix_handle_t smhandle) {
+    auto buffer_ptr =
+        reinterpret_cast<sycl::buffer<T, 1> *>(smhandle->row_container.buffer_ptr.get());
+    return buffer_ptr->template get_access<sycl::access::mode::read_write>(cgh);
+}
+
+template <typename T>
+auto get_col_accessor(sycl::handler &cgh, matrix_handle_t smhandle) {
+    auto buffer_ptr =
+        reinterpret_cast<sycl::buffer<T, 1> *>(smhandle->col_container.buffer_ptr.get());
+    return buffer_ptr->template get_access<sycl::access::mode::read_write>(cgh);
+}
+
+template <typename T>
+auto get_int_accessors(sycl::handler &cgh, matrix_handle_t smhandle) {
+    return std::array<sycl::accessor<T, 1>, 2>{ get_row_accessor<T>(cgh, smhandle),
+                                                get_col_accessor<T>(cgh, smhandle) };
+}
+
+template <typename Functor, typename... CaptureOnlyAcc>
+void submit_host_task(sycl::handler &cgh, sycl::queue &queue, Functor functor,
+                      CaptureOnlyAcc... capture_only_accessors) {
+    // Only capture the accessors to ensure the dependencies are properly handled
+    // The accessors's pointer have already been set to the native container types in previous functions
+    cgh.host_task([functor, queue, capture_only_accessors...](sycl::interop_handle ih) {
```

Comment:
Can you support the new SYCL_EXT_ONEAPI_ENQUEUE_NATIVE_COMMAND extension for all host_task calls that are calling native `Cuspare*` as it is done here https://github.com/oneapi-src/oneMKL/compare/develop...hjabird:oneMKL:hjab/acpp_enqueue_custom_operation via e.g.

```suggestion
    #ifdef SYCL_EXT_ONEAPI_ENQUEUE_NATIVE_COMMAND
    cgh.ext_codeplay_enqueue_native_command([functor, queue, capture_only_accessors...](sycl::interop_handle ih){
#else
    cgh.host_task([functor, queue, capture_only_accessors...](sycl::interop_handle ih) {
#endif
```

?

Reply:
Yes, I started working on this change. I haven't gotten around to test it yet, I'm working on some other changes as well.

Reply:
The tip of the branch now supports the `enqueue_native_command` extension. There are some caveats, as far I can tell the extension can't be used with `*_buffer_size` or `*_optimize` functions as they are synchronous in cuSPARSE and may not enqueue anything to the queue. The compute step is using the extension but it needs some explicit waits to ensure the optimize step as completed before the computation is scheduled to the CUDA stream. In short I'm not sure if this extension will be of any use for the SPARSE domain but it is worth trying it.

Reply:
Resolving this comment as I think this is answered. Feel free to unresolve if it is not.